### PR TITLE
YOLOv5 recipes and apply a recipe tutorial

### PR DESCRIPTION
**Pull request by markurtz**

### CODE REVIEW

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+Your goal after this is to create a smaller, faster model that recovers to the pre-trained baseline.
+
+Creating a pre-trained model involves two steps: 1) setting up the data and 2) training the model.
+
+**Note**: If using your custom data, the Ultralytics repo contains a walk-through for [training custom data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data). 
+Otherwise, setup scripts for both [VOC](https://cs.stanford.edu/~roozbeh/pascal-context/) and [COCO](https://cocodataset.org/#home) can be found under the [yolov5/data/scripts path](https://github.com/ultralytics/yolov5/tree/master/data/scripts).
+
+### Setting Up the Data
+
+1. For this tutorial, we run the COCO setup script with the following command from the root of the yolov5 repository:
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:39]:
```suggestion
1. For this tutorial, run the COCO setup script with the following command from the root of the `yolov5` repository:
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+   
+    You are ready to train the model.
+
+### Training the Model
+
+The training command will take multiple hours to complete since it is training from scratch (anywhere from 12 hours for YOLOV5s to 48 hours for YOLOv5l on 4 A100's). 
+Afterward, you will have a model that achieves roughly 0.556 mAP@0.5 or 0.654 mAP@0.5 on the COCO dataset for YOLOv5s and YOLOv5l respectively ready for sparsifying.
+
+1. To create a deployable model, you will train a YOLOv5 version with HardSwish activations from scratch.
+   To decide which version to use, consult the table given in the later [Applying a Recipe](#applying-a-recipe) section.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:82]:
```suggestion
   To decide which version to use, consult the table below in the [Applying a Recipe](#applying-a-recipe) section. 
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+   To decide which version to use, consult the table given in the later [Applying a Recipe](#applying-a-recipe) section. 
+   The training commands for both are given below.
+   Run the commands from the root of the `yolov5` repository folder (`cd integrations/ultralytics-yolov5/yolov5`).
+   
+   YOLOv5s:
+   ```bash
+   python train.py --cfg ../models/yolov5s.yaml --weights "" --data coco.yaml --hyp data/hyp.scratch.yaml
+   ```    
+
+   YOLOv5L:
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:91]:
```suggestion
   YOLOv5l:
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+   ```bash
+   python train.py --cfg ../models/yolov5l.yaml --weights "" --data coco.yaml --hyp data/hyp.scratch.yaml
+   ```
+2. Validate that the training commands completed successfully by checking under the newly created `runs/train/exp` path for the trained weights.
+   The best trained weights will be found at `runs/train/exp/weights/best.pt` and will be used later for further sparsification.
+   The full results are also visible in this folder under multiple formats including tensorboard runs, txt files, and png files.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:97]:
```suggestion
   The full results are also visible in this folder under multiple formats including TensorBoard runs, txt files, and png files.
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    | [YOLOv5l Pruned Quantized](https://github.com/neuralmagic/sparseml/blob/main/integrations/ultralytics-yolov5/recipes/yolov5l.pruned_quantized.md)        | Creates a highly sparse, INT8 YOLOv5l model that recovers reasonably close to the baseline model.                               | ``` python train.py --cfg ../models/yolov5l.yaml --weights PATH_TO_COCO_PRETRAINED_WEIGHTS --data coco.yaml --hyp data/hyp.scratch.yaml --recipe ../recipes/yolov5l.pruned_quantized.md ``` | 0.525        | --- MB       | --- img/sec              |
+
+    ** DeepSparse Performance measured on an AWS C5 instance with 24 cores, batch size 64, and 640x640 input with version 1.6 of the DeepSparse Engine.
+
+2. Select a recipe to use on top of the pretrained model you created.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:146]:
```suggestion
2. Select a recipe to use on top of the pre-trained model you created.
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+    ***Notes About Recipe Selection***
+    - The recipes are not interchangeable between model versions. For example, if you created a pretrained YOLOv5l model then be sure to use a recipe listed for v5l.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:149]:
```suggestion
    - The recipes are not interchangeable between model versions. For example, if you created a pre-trained YOLOv5l model then be sure to use a recipe listed for v5l.
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    - If your hardware does not support quantized networks for inference speedup or complete recovery is very important, then Neural Magic recommends using the  `pruned` recipe. The recipe to use depends on how long you are willing to train and how vital full recovery is. Consult the table above for this comparison.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:150]:
```suggestion
    - If your hardware does not support quantized networks for inference speedup or complete recovery is very important, then Neural Magic recommends using the  `pruned` recipe. The recipe to use depends on how long you are willing to train and how vital full recovery is. Consult the table above for this comparison. Review `/proc/cpuinfo` for the flag `avx512_vnni` to verify chipset compatibility.
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+   python train.py --cfg ../models/yolov5s.yaml --weights PATH_TO_COCO_PRETRAINED_WEIGHTS --data coco.yaml --hyp data/hyp.scratch.yaml --recipe PATH_TO_SPARSIFICATION_RECIPE
+   ```
+    After applying a recipe, you are ready to export for inference.
+
+## Exporting for Inference
+
+This step loads a checkpoint file of the best weights measured on the validation set, and converts it into the more common inference formats. 
+Then, you can run the file through a compression algorithm to reduce its deployment size and run it in an inference engine such as [DeepSparse](https://github.com/neuralmagic/deepsparse).
+
+When you applied a recipe in the previous step, the sparsification run created a new `exp` directory under the yolov5 runs directory:
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:166]:
```suggestion
When you applied a recipe in the previous step, the sparsification run created a new `exp` directory under the `yolov5/runs` directory:
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+`-- setup_integration.sh
+```
+
+The `best.pt` file contains a checkpoint of the best weights measured on the validation set.
+These weights can be loaded into the `train.py` and `test.py` scripts now. 
+However, other formats are generally more friendly for other inference deployment platforms, such as [ONNX](https://onnx.ai/).
+
+The [`export.py` script](https://github.com/neuralmagic/yolov5/blob/master/models/export.py) handles the logic behind loading the checkpoint and converting it into the more common inference formats, as described here.
+
+1. Enter the following command to load the PyTorch graph, convert to ONNX, and then correct any misformatted pieces for the pruned and quantized models.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:202]: Will it be obvious to the user what "misformatted pieces" are? Should you provide an examples clause....e.g., such as x, y, or z? Or not needed and it will be quite obvious?

Comment (rahul-tuli) [reply integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:202]: I think It should be pretty obvious, will let @markurtz take the call

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md (hunk 7 above)
Comment (markurtz) [reply integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:150]: Let's default them to using the deepsparse API for this detection rather than trying to parse the /proc/cpuinfo. Will update with this change

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md (hunk 9 above)
Comment (markurtz) [reply integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:202]: This is supposed to be stating that any misformatted pieces for the graph will be corrected by the command. I'll make a few edits to make this more obvious

**Comment by markurtz** [comment]
@jeanniefinks updated for review

**Review by jeanniefinks — DISMISSED** [review]
Fab

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    |-- |-- weights
+    |   |   |-- best.pt
+    |   |   |-- best.onnx
+    |   |   `-- last.pt
+    `-- ...
+    ```
+
+2. Now you can run the `.onnx` file through a compression algorithm to reduce its deployment size and run it in ONNX-compatible inference engines such as [DeepSparse](https://github.com/neuralmagic/deepsparse).
+   The DeepSparse Engine is explicitly coded to support running sparsified models for significant improvements in inference performance. 
+   An example for benchmarking and deploying YOLOv3 models with DeepSparse can be found [here](https://github.com/neuralmagic/deepsparse/tree/main/examples/ultralytics-yolo).
```
Comment (bfineran) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:242]:
```suggestion
   An example for benchmarking and deploying YOLOv5 models with DeepSparse can be found [here](https://github.com/neuralmagic/deepsparse/tree/main/examples/ultralytics-yolo).
```

File: integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+## Wrap-Up
+
+Neural Magic recipes simplify the sparsification process by encoding the hyperparameters and instructions needed to create highly accurate pruned and pruned-quantized YOLOv3 models.
```
Comment (bfineran) [inline integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md:246]:
```suggestion
Neural Magic recipes simplify the sparsification process by encoding the hyperparameters and instructions needed to create highly accurate pruned and pruned-quantized YOLOv5 models. 
```

**Review by bfineran — COMMENTED** [review]
LGTM pending updating straggling v3 references

**Review by bfineran — APPROVED** [review]

**Review by jeanniefinks — APPROVED** [review]
